pkg/querier/handler/handler.go

```diff
@@ -0,0 +1,671 @@
+	}
+
+	ctx := r.Context()
+	if to := r.FormValue("timeout"); to != "" {
+		var cancel context.CancelFunc
+		timeout, err := parseDuration(to)
+		if err != nil {
+			return invalidParamError(err, "timeout")
+		}
+
+		ctx, cancel = context.WithTimeout(ctx, timeout)
+		defer cancel()
+	}
+
+	opts, err := extractQueryOpts(r)
+	if err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+	}
+	qry, err := api.QueryEngine.NewRangeQuery(ctx, api.Queryable, opts, r.FormValue("query"), start, end, step)
+	if err != nil {
+		return invalidParamError(err, "query")
+	}
+	// From now on, we must only return with a finalizer in the result (to
+	// be called by the caller) or call qry.Close ourselves (which is
+	// required in the case of a panic).
+	defer func() {
+		if result.finalizer == nil {
+			qry.Close()
+		}
+	}()
+
+	ctx = httputil.ContextFromRequest(ctx, r)
+
+	res := qry.Exec(ctx)
+	if res.Err != nil {
+		return apiFuncResult{nil, returnAPIError(res.Err), res.Warnings, qry.Close}
+	}
+
+	// Optional stats field in response if parameter "stats" is not empty.
+	sr := api.statsRenderer
+	if sr == nil {
+		sr = defaultStatsRenderer
+	}
+	qs := sr(ctx, qry.Stats(), r.FormValue("stats"))
+
+	return apiFuncResult{createPrometheusResponse(&queryData{
+		ResultType: res.Value.Type(),
+		Result:     res.Value,
+		Stats:      qs,
+	}), nil, res.Warnings, qry.Close}
+}
+
+func returnAPIError(err error) *apiError {
+	if err == nil {
+		return nil
+	}
+
+	cause := errors.Unwrap(err)
+	if cause == nil {
+		cause = err
+	}
+
+	switch cause.(type) {
+	case promql.ErrQueryCanceled:
+		return &apiError{errorCanceled, err}
+	case promql.ErrQueryTimeout:
+		return &apiError{errorTimeout, err}
+	case promql.ErrStorage:
+		return &apiError{errorInternal, err}
+	}
+
+	if errors.Is(err, context.Canceled) {
+		return &apiError{errorCanceled, err}
+	}
+
+	return &apiError{errorExec, err}
+}
+
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC()
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC()
+
+	minTimeFormatted = minTime.Format(time.RFC3339Nano)
+	maxTimeFormatted = maxTime.Format(time.RFC3339Nano)
+)
+
+func (api *API) respond(w http.ResponseWriter, data interface{}, warnings storage.Warnings) {
+	statusMessage := statusSuccess
+	var warningStrings []string
+	for _, warning := range warnings {
+		warningStrings = append(warningStrings, warning.Error())
+	}
+	var b []byte
+	var err error
+	switch data.(type) {
+	case queryrange.PrometheusResponse:
+		prometheusResponse, _ := data.(queryrange.PrometheusResponse)
+		prometheusResponse.Status = string(statusMessage)
+		b, err = proto.Marshal(&prometheusResponse)
+	case instantquery.PrometheusInstantQueryResponse:
+		prometheusInstantQueryResponse, _ := data.(instantquery.PrometheusInstantQueryResponse)
+		prometheusInstantQueryResponse.Status = string(statusMessage)
+		b, err = proto.Marshal(&prometheusInstantQueryResponse)
+	}
```

Comment:
respond() function receives a PrometheusResponse or a PrometheusInstantQueryResponse. The struct type is asserted and marshalled into protobuf here

Reply:
It should be impossible to not get either of them because `respond()` is only called after `query()` or `queryRange()` successfully return data without error, but I added a default with error handling for completion [Done]